internal/icinga2/client_api.go

```diff
@@ -427,26 +438,55 @@ func (client *Client) listenEventStream() error {
 		)
 		switch respT := resp.(type) {
 		case *StateChange:
-			// Only process HARD states
+			// Ignore any state changes while in downtime
+			if respT.DowntimeDepth > 0 {
+				client.Logger.Debugf("Skipping state change, checkable is in downtime, %#v", respT)
+				continue
+			}
+
+			// Only process HARD states.
+			// The checkable might also be in the flapping state, but Icinga 2 doesn't expose "is_flapping" to the
+			// event streams for state change. So, the only thing we can do here is to ignore soft state changes.
 			if respT.StateType == StateTypeSoft {
 				client.Logger.Debugf("Skipping SOFT State Change, %#v", respT)
 				continue
 			}
 
 			ev, err = client.buildHostServiceEvent(client.Ctx, respT.CheckResult, respT.State, respT.Host, respT.Service)
 			evTime = respT.Timestamp.Time()
-
 		case *AcknowledgementSet:
-			ev, err = client.buildAcknowledgementEvent(client.Ctx, respT.Host, respT.Service, respT.Author, respT.Comment)
+			ev, err = client.buildAcknowledgementEvent(client.Ctx, respT.Host, respT.Service, respT.Author, respT.Comment, false)
+			evTime = respT.Timestamp.Time()
+		case *AcknowledgementCleared:
+			ev, err = client.buildAcknowledgementEvent(client.Ctx, respT.Host, respT.Service, "", "", true)
 			evTime = respT.Timestamp.Time()
-
-		// case *AcknowledgementCleared:
 		// case *CommentAdded:
 		// case *CommentRemoved:
 		// case *DowntimeAdded:
-		// case *DowntimeRemoved:
-		// case *DowntimeStarted:
-		// case *DowntimeTriggered:
+		case *DowntimeRemoved:
+			ev, err = client.buildDowntimeEvent(client.Ctx, respT.Downtime, false)
+			evTime = respT.Timestamp.Time()
+		case *DowntimeStarted:
+			if !respT.Downtime.IsFixed {
+				// This may never happen, but Icinga 2 does the same thing, and we need to ignore the start
+				// event for flexible downtime, as there will definitely be a triggered event for it.
```

Comment:
As it "may never happen", a debug log message might be a good idea to catch exactly those bugs one thinks they shouldn't be possible to exist.